### PR TITLE
Initial spot instance support

### LINKED ISSUE #30: Support for spot instances
In the todo at the top of the flintrock script, we see:
```   
* Support for spot instances.
        - Show wait reason (capcity oversubscribed, price too low, etc.).
```

Currently it seems like a lot of the machinery for spot instances is in place, including threading the requested spot price through to `launch_ec2`. Would you accept a PR that actually made this live? the spark_ec2.py way is to launch the root node as non-spot and the workers as spot instances -- would this work for a first version?

### PR DESCRIPTION
**Pull request by ericmjonas**
This isn't a final pull request but more a point to begin discussion. It works successfully in `us-west-2` and `us-east-1`. 

When you make a spot instance request for N machines, this might result in up to N "reservations" for some number of machines (such that the total is the size of your spot request). Hence having to aggregate the results a bit - both types now contribute to an `all_instances` variable. 

@nchammas is this what you were thinking? Any feedback? 

Fixes #30

### CODE REVIEW

**Comment by nchammas** [comment]
The general approach here seems fine to me. I'll go through and give specific feedback.

File: flintrock.py [hunk 1]
```diff
@@ -526,42 +526,104 @@ def launch_ec2(
...
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized,
+                
+                )
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+            # now we loop:
+            print("Waiting for spot instances to be granted...")
+            try:
+                while True:
```
Comment (nchammas) [inline flintrock.py]: Is there any way to simplify this block of code? It's a bit hard to follow. Can we reuse, for example, [this pattern](https://github.com/nchammas/flintrock/blob/a1a5a7e8216b61eeaf8f33512c3da2b5142c542d/flintrock.py#L559-L568) for waiting on resources to reach a certain state?

File: flintrock.py [hunk 2]
```diff
@@ -526,42 +526,104 @@ def launch_ec2(
...
+            except:
+                print("Canceling spot instance requests")
+                connection.cancel_spot_instance_requests(slave_reqs)
+                # Log a warning if any of these requests actually launched instances:
+                # (master_nodes, slave_nodes) = get_existing_cluster(
+                #     conn, opts, cluster_name, die_on_error=False)
+                # running = len(master_nodes) + len(slave_nodes)
+                # if running:
+                #     print(("WARNING: %d instances are still running" % running), file=stderr)
+                sys.exit(0)
```
Comment (nchammas) [inline flintrock.py]: Interrupted commands should `sys.exit(1)`.

File: flintrock.py [hunk 3]
```diff
@@ -526,42 +526,104 @@ def launch_ec2(
...
+                        print("All {machine_num} granted".format(machine_num = machine_num))
+                        reservation_resultset = connection.get_all_reservations(active_instance_ids)
+
+                        all_instances = []
+                        for reservation in reservation_resultset:
+                            all_instances += reservation.instances
+                            
+                        break
+                    else:
+                        print("{grant} of {req} instances granted, waiting longer".format(
```
Comment (nchammas) [inline flintrock.py]: Would this be a good place to show the [bid status](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html)? We don't have to do it in this PR, but I just want to mention it.

I've been bitten in the past while using spark-ec2 by waiting unknowingly for requests that typically don't get fulfilled, e.g. due to `capacity-oversubscribed` or `price-too-low`.

File: flintrock.py [hunk 4]
```diff
@@ -526,42 +526,104 @@ def launch_ec2(
         vpc_id=vpc_id,
         region=region)
 
+    machine_num = num_slaves + 1
```
Comment (nchammas) [inline flintrock.py]: Nitpick: For consistency, we should probably call this `num_instances`.

File: flintrock.py [hunk 5]
```diff
@@ -526,42 +526,104 @@ def launch_ec2(
...
-
-        time.sleep(10)  # AWS metadata eventual consistency tax.
-
-        # TODO: Move this to a reusable function and add a limit on
-        #       wait time.
-        while True:
-            for instance in reservation.instances:
-                if instance.state == 'running':
-                    continue
+        #spot instance request
```
Comment (nchammas) [inline flintrock.py]: Nitpick: I'd generally prefer not to have comments explaining something if it's straightforward.

File: flintrock.py [hunk 6]
```diff
@@ -573,7 +635,7 @@ def launch_ec2(
             tags={
                 'flintrock-role': 'slave',
                 'Name': '{c}-slave'.format(c=cluster_name)})
-
+
```
Comment (nchammas) [inline flintrock.py]: Style: Blank lines should not have any whitespace.

Travis (which currently just runs `lint.sh`) will flag problems like these for you, so they should be easy to find and fix. You can see this in the PR build status.

**Comment by nchammas** [comment]
Thanks for taking a crack at this! I left some specific feedback. My main concern is about the complexity of the main loop to wait for the spot instances. Otherwise, I think we can get this in for Flintrock's initial release.

By the way, you need to pull the latest changes into your fork. You're missing some [changes I introduced yesterday](https://github.com/nchammas/flintrock/blob/a1a5a7e8216b61eeaf8f33512c3da2b5142c542d/flintrock.py#L516-L519) which you'll need to remove.

File: flintrock.py (hunk 3 above)
Comment (nchammas) [inline flintrock.py]: On second thought, perhaps for starters we should just wait for the spot instances and not print anything here. The repeated "waiting longer" messages could be spammy.

What's your feeling on this?

File: flintrock.py (hunk 1 above)
Comment (ericmjonas) [inline flintrock.py]: I've cleaned up the code considerably in the request code. The challenge is we have a 1 spot request -> many reservations model, and so it's just a bit tricky. See the latest commits to this branch.

File: flintrock.py (hunk 2 above)
Comment (ericmjonas) [inline flintrock.py]: fixed

File: flintrock.py (hunk 3 above)
Comment (ericmjonas) [inline flintrock.py]: I have found that it's nice to have this heartbeat -- sometimes it can take up to 20 min for a large spot instance to be granted, even if you're far above the current market price.

File: flintrock.py (hunk 4 above)
Comment (ericmjonas) [inline flintrock.py]: fixed

File: flintrock.py (hunk 5 above)
Comment (ericmjonas) [inline flintrock.py]: fixed

File: flintrock.py (hunk 6 above)
Comment (ericmjonas) [inline flintrock.py]: Fixed, new pr should work

File: flintrock.py [hunk 7]
```diff
@@ -526,42 +526,98 @@ def launch_ec2(
...
+                instance_type=instance_type,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+
+            print("Waiting for spot instances to be granted...")
```
Comment (nchammas) [inline flintrock.py]: Do we need this second print here if we've already told the user we've requested spot instances? Maybe we should just add some ellipsis to the print in [L534](https://github.com/nchammas/flintrock/pull/35/files#diff-1ec778e3c6cde832d96a7c08ba31f98eR534) to suggest we are waiting and get rid of this print here.

File: flintrock.py [hunk 8]
```diff
@@ -526,42 +526,98 @@ def launch_ec2(
...
+        if spot_price is not None:
+            print("Requesting {c} spot instances at a max price of {p}".format(c=num_instances, p=spot_price))
+
+            instance_reqs = connection.request_spot_instances(
+                price=spot_price,
+                image_id=ami,
+                count=(num_instances),
+                key_name=key_name,
+                instance_type=instance_type,
+                placement=availability_zone,
```
Comment (nchammas) [inline flintrock.py]: I noticed in an earlier version of this PR you were experimenting with the `launch_group` parameter. Can you tell me a bit about that? According to [the docs](http://boto.readthedocs.org/en/latest/ref/ec2.html#boto.ec2.connection.EC2Connection.request_spot_instances), it sounds useful here, but I'm not sure.

File: flintrock.py [hunk 9]
```diff
@@ -526,42 +526,98 @@ def launch_ec2(
...
+            print("Waiting for spot instances to be granted...")
+            try:
+                while True:
+                    time.sleep(30)
+
+                    # Note that there can be outstanding spot instance requests
+                    # that are not part of our current requests (say, from
+                    # another process), and so we have to filter for just
+                    # our requests
+                    all_account_reqs = connection.get_all_spot_instance_requests()
```
Comment (nchammas) [inline flintrock.py]: I think we can avoid some of the filter logic below if we only request the pending spot requests here by passing in [`request_ids`](http://boto.readthedocs.org/en/latest/ref/ec2.html#boto.ec2.connection.EC2Connection.get_all_spot_instance_requests). Here's [an example](http://www.imlucas.com/post/55003108849/waiting-for-spot-instances-to-be-fulfilled-with) of this.

File: flintrock.py (hunk 7 above)
Comment (ericmjonas) [inline flintrock.py]: Great idea, will add

File: flintrock.py [hunk 10]
```diff
@@ -526,42 +526,98 @@ def launch_ec2(
...
+                placement_group=placement_group,
+                tenancy=tenancy,
+                ebs_optimized=ebs_optimized,
+                instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
+
+            time.sleep(10)  # AWS metadata eventual consistency tax.
+
+            # TODO: Move this to a reusable function and add a limit on
+            #       wait time.
+            while True:
```
Comment (nchammas) [inline flintrock.py]: Do we not need this block to wait for the instances to come up if we are requesting spot instances?

File: flintrock.py (hunk 8 above)
Comment (ericmjonas) [inline flintrock.py]: We basically never use launch groups -- I was just following what spark_ec2 was doing but then cleaned up the parts that didn't seem mission-critical at the time. There are a lot of situations where having a few of your spot instances die is ok for a running spark job.

File: flintrock.py (hunk 9 above)
Comment (ericmjonas) [inline flintrock.py]: Great find!

File: flintrock.py (hunk 10 above)
Comment (ericmjonas) [inline flintrock.py]: I think we already do the appropriate (consistency) wait in the spot-instance request code. After our initial request, we wait 30s for the spot instances to be fulfilled.

Comment (nchammas) [inline flintrock.py]: I mean the wait for the instances to get to a `running` state. We do that so the instances are guaranteed to have been assigned IP addresses, which the rest of the launch code will expect.

Right now we don't explicitly wait for the instances be `running`, so I'm wondering if we should move this block of code outside the `else` clause. Does that make sense?

**Comment by nchammas** [comment]
Thanks for these latest changes @ericmjonas! The PR looks much better. I went through again and left more comments, which you are already responding to (sweet!). Basically, I still think we can greatly simplify the logic for waiting on the spot requests, but otherwise this is shaping up nicely.

A few notes to myself for the future:
* When we upgrade to boto3 (#22) we can probably replace the custom spot request wait logic entirely with [`EC2.Waiter.SpotInstanceRequestFulfilled`](http://boto3.readthedocs.org/en/latest/reference/services/ec2.html#EC2.Waiter.SpotInstanceRequestFulfilled).
* It would be nice in the future to replace the repeated print statements in the spot request wait loop with a progress bar (like proposed in #27) or a simple spinner.

File: flintrock.py (hunk 8 above)
Comment (nchammas) [inline flintrock.py]: Okie doke, makes sense.

**Comment by ericmjonas** [comment]
@nchammas I think I messed up on my attempt to rebase to your master and now for some reason my tree is really ugly. I could try fixing but I'm guessing that we'll just squash the entire commit hist for the eventual final PR?

**Comment by nchammas** [comment]
Yeah, before merging the PR I'll ask you to squash it into one commit. Don't worry about it for now.

By the way, the way I've pulled in changes to personal forks in the past is simple. If you cloned your fork locally using GitHub, you should have a remote already setup for you called `nchammas`.

```sh
# from your fork, on master
git remote -v
git pull nchammas master
# switch to your development branch, then
git rebase master
```

File: flintrock.py [hunk 11]
```diff
@@ -533,31 +528,82 @@ def launch_ec2(
...
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+
+            try:
+                while True:
+                    time.sleep(30)
```
Comment (nchammas) [inline flintrock.py]: I haven't launched spot instances in a while. Is this typically how long we want to wait between requests? Seems a bit long, maybe it's appropriate.

File: flintrock.py [hunk 12]
```diff
@@ -533,31 +528,82 @@ def launch_ec2(
...
+                        reservation_resultset = connection.get_all_reservations(active_instance_ids)
+
+                        all_instances = []
+                        for reservation in reservation_resultset:
+                            all_instances += reservation.instances
+                        break
+                    else:
+                        print("{grant} of {req} instances granted, waiting longer".format(
+                            grant=len(active_instance_ids), req=num_instances))
+            except:
```
Comment (nchammas) [inline flintrock.py]: We shouldn't have bare excepts like this.

```python
except Exception as e:
    print(e)
    print("Canceling spot instance requests...")  # added ellipsis here
    ...
```

This is a bit more user-friendly.

File: flintrock.py [hunk 13]
```diff
@@ -533,31 +528,82 @@ def launch_ec2(
...
+                print("Canceling spot instance requests")
+                connection.cancel_spot_instance_requests(instance_reqs)
+                # FIXME add warning if any nodes actually launched instances
```
Comment (nchammas) [inline flintrock.py]: Nit: I've been using `TODO: ...` for notes like this.

File: flintrock.py [hunk 14]
```diff
@@ -533,31 +528,82 @@ def launch_ec2(
...
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+
+            try:
+                while True:
+                    time.sleep(30)
+
+                    reqs = connection.get_all_spot_instance_requests(spot_instance_request_ids)
```
Comment (nchammas) [inline flintrock.py]: This is better, but I was hoping we could really simplify things, like:

```python
pending_request_ids = [req.id for req in instance_reqs]

while pending_request_ids:
    spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
    pending_request_ids = [req.id for req in spot_requests if req.state != 'active']
```

What do you think?

**Comment by nchammas** [comment]
Whoops! Sorry Eric, I missed the fact that you submitted some updates yesterday. I just went through and made another round of comments.

File: flintrock.py (hunk 11 above)
Comment (ericmjonas) [inline flintrock.py]: I think so -- even small spot instance requests (say for 5 instances) can take ~90 seconds to be approved (by whatever method Amazon uses to make sure you're not going to break everything) and then 2-3 min to grant, depending on price.

**Comment by nchammas** [comment]
I'm having a little trouble applying this patch locally so I can test it out. Is it working for you in this state?

File: flintrock.py [hunk 15]
```diff
@@ -533,31 +528,80 @@ def launch_ec2(
...
+                    spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
+                    pending_request_ids = [req.id for req in spot_requests if req.state != 'active']
 
-        time.sleep(10)  # AWS metadata eventual consistency tax.
+                    if len(pending_request_ids) > 0:
+                        print("{grant} of {req} instances granted, waiting longer".format(
+                            grant=num_instances - len(pending_request_ids), req=num_instances))
+
+                print("All {num_instances} granted".format(num_instances=num_instances))
+                spot_requests = connection.get_all_spot_instance_requests(pending_request_ids)
```
Comment (nchammas) [inline flintrock.py]: If I'm not mistaken, `pending_request_ids` will be empty here because they are all fulfilled, so this won't work. You probably want to reuse `instance_reqs` from earlier, no?

Comment (ericmjonas) [inline flintrock.py]: I'm sorry, I'm not sure how that made it through; fixing now.

**Comment by nchammas** [comment]
Can you add "Fixes #30." to the PR description, by the way? That way when we merge this PR in GitHub will automatically close the issue for us.

**Comment by nchammas** [comment]
@ericmjonas, if you're ready, I think now's a good time to squash these commits into one. If you need help doing that, pop in to the [Flintrock Gitter room](https://gitter.im/nchammas/flintrock) and I'll help you out. Once we have that done I think I can merge this in!

File: flintrock.py [hunk 16]
```diff
@@ -533,31 +528,82 @@ def launch_ec2(
...
+                reservation_resultset = connection.get_all_reservations(instance_ids)
+
+                all_instances = []
+                for reservation in reservation_resultset:
+                    all_instances += reservation.instances
+
+            except Exception as e:
+                print(e)
+                print("Canceling spot instance requests...")
+                connection.cancel_spot_instance_requests(instance_reqs)
```
Comment (nchammas) [inline flintrock.py]: Here you want `pending_request_ids`.

**Comment by nchammas** [comment]
Alright, I tested this out and it looks good to me. There's an issue with the exception handling, but that's OK. I'll pull your changes in to a new PR and clean that up along with some minor style issues and get this merged in.

Thanks for working on this!
